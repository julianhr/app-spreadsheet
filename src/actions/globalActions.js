import { createAction } from 'redux-starter-kit'


export function setTableDimensions(rows, columns) {
  return {
    type: 'SET_TABLE_DIMENSIONS',
    payload: {
      rows,
      columns,
    }
  }
}

export function setInputterValueEvent(value, cursorPos) {
  return {
    type: 'SET_INPUTTER_VALUE_EVENT',
    payload: {
      value,
      cursorPos: cursorPos === undefined
        ? value.length
        : cursorPos
    }
  }
}

export function setActiveCell(location) {
  return (dispatch, getState) => {
    const cellData = getState().tableData[location]
    const entered = (cellData || {}).entered || ''

    dispatch(setInputterValueEvent(entered, entered.length))

    return Promise.resolve(
      dispatch({
        type: 'SET_ACTIVE_CELL',
        payload: {
          entered,
          location,
        }
      })
    )
  }
}

export function setFloatingInputterInteractive(isInteractive) {
  return {
    type: 'SET_FLOATING_INPUTTER_INTERACTIVE',
    payload: isInteractive
  }
}

export function openFloatingInputter(cellRect, willReplaceValue, isInteractive) {
  return (dispatch, getState) => {
    const state = getState()
    const value = willReplaceValue ? '' : state.global.activeCell.entered

    dispatch(setInputterValueEvent(value))

    return Promise.resolve(
      dispatch({
        type: 'OPEN_FLOATING_INPUTTER',
        payload: {
          cellRect,
          isOpen: true,
          isInteractive,
          willOpen: false,
        }
      })
    )
  }
}

export function resetInputterValueEvent() {
  return (dispatch, getState) => {
    const location = getState().global.activeCell.location
    const cellData = getState().tableData[location]
    const entered = (cellData || {}).entered || ''

    dispatch(setInputterValueEvent(entered))
  }
}

export function scheduleFloatingInputter(isInteractive) {
  return (dispatch, getState) => {
    const { floatingInputter } = getState().global

    if (floatingInputter.isOpen && floatingInputter.isInteractive === isInteractive) {
      return
    }

    return Promise.resolve(
      dispatch({
        type: 'SCHEDULE_FLOATING_INPUTTER',
        payload: {
          isInteractive
        },
      })
    )
  }
}

export function unscheduleFloatingInputter() {
  return {
    type: 'UNSCHEDULE_FLOATING_INPUTTER',
    payload: null,
  }
}

export const closeFloatingInputter = createAction('CLOSE_FLOATING_INPUTTER')
