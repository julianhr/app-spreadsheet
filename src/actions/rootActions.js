import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')

export function moveActiveCell(key) {
  return (dispatch, getState) => {
    const { activeCell, rows, columns } = getState()
    const currRow = activeCell.codePointAt(0) - 'a'.codePointAt(0)
    const currCol = Number.parseInt(activeCell[1]) - 1
    let nextRow = currRow
    let nextCol = currCol

    if (key === 'ArrowUp') { nextRow = currRow - 1 }
    else if (key === 'ArrowRight') { nextCol = currCol + 1 }
    else if (key === 'ArrowDown') { nextRow = currRow + 1 }
    else if (key === 'ArrowLeft') { nextCol = currCol - 1 }

    if (Math.min(nextRow, nextCol) >= 0 && nextRow < rows && nextCol < columns) {
      nextRow = String.fromCodePoint('a'.codePointAt(0) + nextRow)
      nextCol = nextCol + 1
      dispatch(setActiveCell(`${nextRow}${nextCol}`))
    }
  }
}
