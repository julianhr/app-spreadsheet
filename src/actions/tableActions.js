import graph from '~/formulas/graph'
import Interpreter from '../formulas/Interpreter'


export function setCellData(location, entered) {
  const interpreter = new Interpreter(location)
  const result = interpreter.interpret(entered)


  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function replaceCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().table[location]

    graph.delVertex(location)

    if (value === undefined) {
      return Promise.resolve()
    }


    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
