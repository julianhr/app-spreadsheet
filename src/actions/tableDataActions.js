export function setCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().tableData[location]

    if (value === undefined) {
      return Promise.resolve()
    }

    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
