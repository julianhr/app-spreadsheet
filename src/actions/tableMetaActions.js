import { clamp } from '~/library/utils'
import {
  MIN_COL_WIDTH,
  MAX_COL_WIDTH,
  MIN_ROW_HEIGHT,
  MAX_ROW_HEIGHT,
} from '~/library/constants'


export function setColWidth(colLabel, width) {
  if (typeof width !== 'number') {
    throw new Error('width is not a number')
  }

  return {
    type: 'SET_COL_WIDTH',
    payload: { colLabel, width: clamp(width, MIN_COL_WIDTH, MAX_COL_WIDTH) }
  }
}

export function setRowHeight(rowLabel, height) {
  if (typeof height !== 'number') {
    throw new Error('height is not a number')
  }

  return {
    type: 'SET_ROW_HEIGHT',
    payload: { rowLabel, height: clamp(height, MIN_ROW_HEIGHT, MAX_ROW_HEIGHT) }
  }
}
