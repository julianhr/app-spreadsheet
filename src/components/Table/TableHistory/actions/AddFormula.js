import ReduxActions from './ReduxActions'


class AddFormula extends ReduxActions {
  constructor(location, oldVertex, newVertex) {
    super()
    this.location = location
    this.entered = newVertex.entered
    this.result = newVertex.result
    this.previous = oldVertex && {
      entered: oldVertex.entered,
      result: oldVertex.result,
    }
  }

  execute() {
    this.reduxConnect.fn.setCellData(this.location, this.entered, this.result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (this.previous) {
      const { entered, result } = this.previous
      this.reduxConnect.fn.setCellData(this.location, entered, result)
    } else {
      this.reduxConnect.fn.clearCellData(this.location)
    }

    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default AddFormula
