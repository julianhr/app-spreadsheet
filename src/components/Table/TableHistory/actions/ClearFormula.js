import ReduxActions from './ReduxActions'


class ClearFormula extends ReduxActions {
  constructor(location) {
    super()
    this.previous = this.getCellData(location)
    this.location = location
  }

  execute() {
    this.reduxConnect.fn.clearCellData(this.location)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (!this.previous) { return }
    const { entered, result } = this.previous
    this.reduxConnect.fn.setCellData(this.location, entered, result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default ClearFormula