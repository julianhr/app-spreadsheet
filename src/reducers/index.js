import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
