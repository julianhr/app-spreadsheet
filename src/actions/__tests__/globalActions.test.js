import { appStoreGen } from '~/reducers/'
import * as actions from '~/actions/globalActions'


describe('globalActions', () => {
  describe('#setActiveCell', () => {
    it('sets currently active cell', async () => {
      const value = 'B-2'
      const appStore = appStoreGen()

      expect(appStore.getState().global.activeCell).toBeNull()
      await appStore.dispatch(actions.setActiveCell(value))
      expect(appStore.getState().global.activeCell).toEqual(value)
    })
  })

  describe('#displayInputter', () => {
    it('sets data for cell inputter', async () => {
      const location = 'B-2'
      const appStore = appStoreGen()

      expect(appStore.getState().global.cellInputter).toBeNull()
      await appStore.dispatch(actions.displayInputter({ testKey: 'test' }))
      expect(appStore.getState().global.cellInputter.testKey).toEqual('test')
    })
  })
})
