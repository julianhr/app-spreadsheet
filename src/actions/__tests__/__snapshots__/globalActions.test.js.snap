// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`globalActions #displayCellInputter sets data for cell inputter 1`] = `
Object {
  "entered": "=5+3",
  "location": "B-2",
}
`;

exports[`globalActions #displayCellInputter sets data for cell inputter 2`] = `
Object {
  "entered": "=5+3",
  "location": "B-2",
}
`;

exports[`globalActions #setActiveCell sets currently active cell 1`] = `
Object {
  "entered": "",
  "location": "B-2",
}
`;
